text/0000-module-resolution.md

```diff
@@ -0,0 +1,340 @@
+
+## Algorithm
+
+Here is the algoritm for module resolution (the instructions are not in a rigurous specification format for brevity and time)
+
+```
+  Let `rootDir` be the initial path for the application, `initModules` 
+  the list of ModuleRecords to be resolved and ModuleRecordEntryList 
+  an empty array.
+  
+  1. Load the LWC configuration file in the rootDir
+    1.1 Search for `lwc.config.js`, if it exist read it.
+    1.2 If lwc.config.js is not found search in package.json
+    (If both are found the configuration file has precedence).
+
+  2. Merge all the ModuleRecords found in the config with 
+  the list of initModules. If an alias with the same name 
+  exist pick initModules over the ones resolved in the config.
+
+  3. For each ModuleRecord:
+
+    3.1  Validate that its a valid ModuleRecord type, skip it otherwise.
+
+    3.2 Resolve ModuleRecord entry:
+
+        3.2.1 If is an AliasModuleRecord validate the path and add create 
+        a ModuleRecordEntry with: 
+          - `specifier` as the `name` value.
+          - `scope` with the closest configuration path.
+          - `entry` with the `path` value.
+
+        3.2.2 If is a DirModuleRecord validate the path and find of the 
+        modules that match the structure:
+        [namespace]/[componentName]/[componentName.{html|css|js|ts}]
+
+        3.2.3 If is a NpmModuleRecord, validate that the package is installed 
+        and GoTo 1. recursively merging all the ModuleRecordEntries.
+
+    3.3 Return all the ModuleRecordEntries collected.
+```
+
+#### Future ModuleRecord types
+
+We will not discard adding another types of module resolutions, like for example, find modules directly from git or even from a url.
+This schema allows us to add 
+
+## Scoping
+
+Its an invariant of the module resolution system to allow for multiple versions of multiple components within a particular application configuration. As the modules are resolved, we keep track of the context on which those modules appear, basically who is the importer (and its path). You can check again the [basic example](#basic_example) on the top.
+
+This concept of scope matches the [current specification of import maps](https://github.com/WICG/import-maps#multiple-versions-of-the-same-module) which will allows us to create the equivalent resolution for the browser.
+
+## Resolving modules at build time.
+
+This algorithm can be used and integrated with tools like Webpack or rollup, in fact we have created both the `@lwc/module-resolver` and the `@lwc/rollup-plugin` packages that we officially support. Other projects in the ecosystem such as `create-lwc-app` also leverage those primitives.
+
+## Alternatives
+
+We discussed different module resolution alteratives previously, but to summarize:
+
+  - We need a custom module resolution to: 
+    - Support the 1:N relationship on packages
+    - Have different build steps
+    - Align with standards in different host environments
```

Comment:
Does NPM have a concept of package "types"? Like a package.json primitive which allows you to say "this has ui components or this has a service" I can see this resolution mechanism becoming a thing

Reply:
No, but thats why we have the `lwc` config key in there.